### PR TITLE
Change how we pick bwc versions to check out

### LINKED ISSUE #44327: Last good commit doesn't work for bwc tests
Also described in #43971 and https://elasticsearch-ci.elastic.co/view/Elasticsearch%20master/job/elastic+elasticsearch+master+release-tests/145/ possibly also failed because of this

### PR DESCRIPTION
**Pull request by alpar-t**
Prior to this PR we always checked out the latest bwc branches and had
an external mechanism to store the bwc versions used for every CI run so
we could both reproduce those builds and run additional tests using the
same combination.

This adds complexities in setting up and maintaining CI and makes it
difficult to set up multi jobs.

This change replaces that mechanism with a time based approach
that looks at the commit date of the current revision and picks the
newest on the bwc branch that's still older than that.
It also makes sure there are no merge commits in this interval.

This new behavior will is ment to be enabled in CI only, for everything
except PR checks that will still use last available bwc revision.

Closes #44327 relates to #43971

### CODE REVIEW

**Comment by atorok** [comment]
I looked at some builds and how they compare to this approach: 

Job building 40490224762dccd7e48a85c5ea8c51d4bfc09bc4 checked out `7.x`at 872ae4d6c26645e6d395a74daef6de0d5f7ba5f2         
which was committed ~ 2 minutes after wereas the current method would have picked e708710f3857995130c50975217824fb5e5d6538
that's the commit right before it.

An older build on e765a00b683b50e81b80a16df8c9ced949461e2e picked ae5c01e2d240b0b6cb76c23ebcc26ed4d9b57b1c for `7.x`.
This approach would have picked 8a6675b9949e18e50193ab9345ced7cd2384a7db. The one the build picked is in fact close to
45 minutes older than the one building. We probably had some queuing in Jenkins.

A third example was on e5805c961e3f220af399713a85ba484966da76ed and picked d78ecd9c096fd7dad9cc12d487f7e4a8a8687868 which is
arround 3 minutes older the new approach would have picked 925bd4faa11a5fbfefe064665b9e5747e26f3f7f which is 2 commits older.

File: distribution/bwc/build.gradle
Comment (mark-vieira) [inline distribution/bwc/build.gradle]: Ideally, what we would doo in this scenario is simply back up to the most recent commit before the merge but we can play it by ear. I suspect we won't encounter this too often, and if we do, we just have to wait until the next non-merge commit.

File: distribution/bwc/build.gradle
Comment (mark-vieira) [inline distribution/bwc/build.gradle]: So when a merge is done w/o a merge commit the merged commit dates are the "push" date, yes?

**Review by mark-vieira — APPROVED** [review]
Couple of questions but otherwise this looks good and should solve a lot of problems.

File: distribution/bwc/build.gradle
Comment (rjernst) [reply distribution/bwc/build.gradle]: Looking back over the last 6 months, I see 2 merge commits, one in particular in January was a manually merged feature branch. While we should not be doing that, we don't have any protections against them appear in the history right now, and failing the build on them existing would be an immediate breakage in CI that could not be fixed.

I suggested we use `--first-parent` in the rev-list command for traversing history. This will ensure we stay on the branch and don't fall into any commits that were made on a feature branch should another commit like that be added. We should then not need to check for merges.

File: distribution/bwc/build.gradle
Comment (atorok) [reply distribution/bwc/build.gradle]: Yes, squash merging create a new commit that gets a date and is pushed to the branch immediately if done trough the GH UI. If it's not done from GH UI, `git pull --rebase` creates a new commit that gets a new date and you can't push the old one without pulling from the branch first.  If one doesn't pass `--rebase` a merge commit would be created.

File: distribution/bwc/build.gradle
Comment (atorok) [reply distribution/bwc/build.gradle]: I don't have a string preference on the matter. 
We don't have any protections against it, but this could at least act as a check instead of making it be lenient. 
We check for merge commits only __after__ the time of the current commit, so creating an empty non merge commit after the merge commit would solve the problem of CI failing leaving the failed job to raise awareness.

Comment (jasontedor) [reply distribution/bwc/build.gradle]: Merge commits for feature branches are acceptable, so that we don't lose history from long-running feature branches worked on by multiple developers (think: CCR, ILM, Zen 2).

Comment (atorok) [reply distribution/bwc/build.gradle]: @jasontedor allowing them in this scenario could mean that we end up picking a commit form within the feature branch. If we just walk the first parent, we'll not have the feature branch in bwc but likely have it in master. In either case this will lead to a more obscure failure. I think I prefer the explicit failure here on the rare occasions, especially that all it takes is to have one commit after it, so one could push an empty commit after the merge commit ( with a lather date ) to avoid the fallout from CI

**Pull request merged**
